File: code/go/ch17x/src/luago/binchunk/reader.go
package binchunk

import "encoding/binary"
import "math"

type reader struct {
	data []byte
}

func (self *reader) readByte() byte {
	b := self.data[0]
	self.data = self.data[1:]
	return b
}

func (self *reader) readBytes(n uint) []byte {
	bytes := self.data[:n]
	self.data = self.data[n:]
	return bytes
}

func (self *reader) readUint32() uint32 {
	i := binary.LittleEndian.Uint32(self.data)
	self.data = self.data[4:]
	return i
}

func (self *reader) readUint64() uint64 {
	i := binary.LittleEndian.Uint64(self.data)
	self.data = self.data[8:]
	return i
}

func (self *reader) readLuaInteger() int64 {
	return int64(self.readUint64())
}

func (self *reader) readLuaNumber() float64 {
	return math.Float64frombits(self.readUint64())
}

func (self *reader) readString() string {
	size := uint(self.readByte())
	if size == 0 {
		return ""
	}
	if size == 0xFF {
		size = uint(self.readUint64()) // size_t
	}
	bytes := self.readBytes(size - 1)
	return string(bytes) // todo
}

func (self *reader) checkHeader() {
	if string(self.readBytes(4)) != LUA_SIGNATURE {
		panic("not a precompiled chunk!")
	}
	if self.readByte() != LUAC_VERSION {
		panic("version mismatch!")
	}
	if self.readByte() != LUAC_FORMAT {
		panic("format mismatch!")
	}
	if string(self.readBytes(6)) != LUAC_DATA {
		panic("corrupted!")
	}
	if self.readByte() != CINT_SIZE {
		panic("int size mismatch!")
	}
	if self.readByte() != CSIZET_SIZE {
		panic("size_t size mismatch!")
	}
	if self.readByte() != INSTRUCTION_SIZE {
		panic("instruction size mismatch!")
	}
	if self.readByte() != LUA_INTEGER_SIZE {
		panic("lua_Integer size mismatch!")
	}
	if self.readByte() != LUA_NUMBER_SIZE {
		panic("lua_Number size mismatch!")
	}
	if self.readLuaInteger() != LUAC_INT {
		panic("endianness mismatch!")
	}
	if self.readLuaNumber() != LUAC_NUM {
		panic("float format mismatch!")
	}
}

func (self *reader) readProto(parentSource string) *Prototype {
	source := self.readString()
	if source == "" {
		source = parentSource
	}
	return &Prototype{
		Source:          source,
		LineDefined:     self.readUint32(),
		LastLineDefined: self.readUint32(),
		NumParams:       self.readByte(),
		IsVararg:        self.readByte(),
		MaxStackSize:    self.readByte(),
		Code:            self.readCode(),
		Constants:       self.readConstants(),
		Upvalues:        self.readUpvalues(),
		Protos:          self.readProtos(source),
		LineInfo:        self.readLineInfo(),
		LocVars:         self.readLocVars(),
		UpvalueNames:    self.readUpvalueNames(),
	}
}

func (self *reader) readCode() []uint32 {
	code := make([]uint32, self.readUint32())
	for i := range code {
		code[i] = self.readUint32()
	}
	return code
}

func (self *reader) readConstants() []interface{} {
	constants := make([]interface{}, self.readUint32())
	for i := range constants {
		constants[i] = self.readConstant()
	}
	return constants
}

func (self *reader) readConstant() interface{} {
	switch self.readByte() {
	case TAG_NIL:
		return nil
	case TAG_BOOLEAN:
		return self.readByte() != 0
	case TAG_INTEGER:
		return self.readLuaInteger()
	case TAG_NUMBER:
		return self.readLuaNumber()
	case TAG_SHORT_STR, TAG_LONG_STR:
		return self.readString()
	default:
		panic("corrupted!") // todo
	}
}

func (self *reader) readUpvalues() []Upvalue {
	upvalues := make([]Upvalue, self.readUint32())
	for i := range upvalues {
		upvalues[i] = Upvalue{
			Instack: self.readByte(),
			Idx:     self.readByte(),
		}
	}
	return upvalues
}

func (self *reader) readProtos(parentSource string) []*Prototype {
	protos := make([]*Prototype, self.readUint32())
	for i := range protos {
		protos[i] = self.readProto(parentSource)
	}
	return protos
}

func (self *reader) readLineInfo() []uint32 {
	lineInfo := make([]uint32, self.readUint32())
	for i := range lineInfo {
		lineInfo[i] = self.readUint32()
	}
	return lineInfo
}

func (self *reader) readLocVars() []LocVar {
	locVars := make([]LocVar, self.readUint32())
	for i := range locVars {
		locVars[i] = LocVar{
			VarName: self.readString(),
			StartPC: self.readUint32(),
			EndPC:   self.readUint32(),
		}
	}
	return locVars
}

func (self *reader) readUpvalueNames() []string {
	names := make([]string, self.readUint32())
	for i := range names {
		names[i] = self.readString()
	}
	return names
}

File: code/csharp/ch13/luavm/binchunk/Reader.cs
using System;
using System.Linq;
using luavm.util;

namespace luavm.binchunk
{
    internal class Reader
    {
        public byte[] Data;

        public byte ReadByte()
        {
            var b = Data[0];
            Data = Data.Skip(1).ToArray();
            return b;
        }

        private uint ReadUint32()
        {
            var bytes = new byte[4];
            Array.ConstrainedCopy(Data, 0, bytes, 0, 4);
            Data = Data.Skip(4).ToArray();
            if (!BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }

            return BitConverter.ToUInt32(bytes, 0);
        }

        private long ReadLuaInteger()
        {
            return (long) ReadUint64();
        }

        private double ReadLuaNumber()
        {
            return BitConverter.Int64BitsToDouble((long) ReadUint64());
        }

        private ulong ReadUint64()
        {
            var bytes = new byte[8];
            Array.ConstrainedCopy(Data, 0, bytes, 0, 8);
            Data = Data.Skip(8).ToArray();
            if (!BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }

            return BitConverter.ToUInt64(bytes, 0);
        }

        private byte[] ReadBytes(uint n)
        {
            var bytes = new byte[n];
            Array.ConstrainedCopy(Data, 0, bytes, 0, (int) n);
            Data = Data.Skip((int) n).ToArray();
            return bytes;
        }

        public void CheckHeader()
        {
//            Console.WriteLine(Encoding.Default.GetString(readBytes(4)));
            if (ConvertUtil.Bytes2String(ReadBytes(4)) != BinaryChunk.LUA_SIGNATURE)
            {
                throw new Exception("not a precompiled chunk!");
            }

            if (ReadByte() != BinaryChunk.LUAC_VERSION)
            {
                throw new Exception("version mismatch!");
            }

            if (ReadByte() != BinaryChunk.LUAC_FORMAT)
            {
                throw new Exception("format mismatch!");
            }

            if (ConvertUtil.Bytes2String(ReadBytes(6)) != BinaryChunk.LUAC_DATA)
            {
                throw new Exception("corrupted!");
            }

            if (ReadByte() != BinaryChunk.CINT_SIZE)
            {
                throw new Exception("int size mismatch!");
            }

            var b = ReadByte();
            if (b != BinaryChunk.CSIZET_SIZE_32 && b != BinaryChunk.CSIZET_SIZE_64)
            {
                throw new Exception("size_t size mismatch!");
            }

            if (ReadByte() != BinaryChunk.INSTRUCTION_SIZE)
            {
                throw new Exception("instruction size mismatch!");
            }

            if (ReadByte() != BinaryChunk.LUA_INTEGER_SIZE)
            {
                throw new Exception("lua_Integer size mismatch!");
            }

            if (ReadByte() != BinaryChunk.LUA_NUMBER_SIZE)
            {
                throw new Exception("lua_Number size mismatch!");
            }

            if (ReadLuaInteger() != BinaryChunk.LUAC_INT)
            {
                throw new Exception("endianness mismatch!");
            }

            if (!ReadLuaNumber().Equals(BinaryChunk.LUAC_NUM))
            {
                throw new Exception("float format mismatch!");
            }
        }

        public Prototype ReadProto(string parentSource)
        {
            var source = ReadString();
            if (source == "")
            {
                source = parentSource;
            }

            return new Prototype
            {
                Source = source,
                LineDefined = ReadUint32(),
                LastLineDefined = ReadUint32(),
                NumParams = ReadByte(),
                IsVararg = ReadByte(),
                MaxStackSize = ReadByte(),
                Code = ReadCode(),
                Constants = ReadConstants(),
                Upvalues = ReadUpvalues(),
                Protos = ReadProtos(source),
                LineInfo = ReadLineInfo(),
                LocVars = ReadLocVars(),
                UpvalueNames = ReadUpvalueNames()
            };
        }

        private string[] ReadUpvalueNames()
        {
            var names = new string[ReadUint32()];
            for (var i = 0; i < names.Length; i++)
            {
                names[i] = ReadString();
            }

            return names;
        }

        private LocVar[] ReadLocVars()
        {
            var locVars = new LocVar[ReadUint32()];
            for (var i = 0; i < locVars.Length; i++)
            {
                locVars[i] = new LocVar
                {
                    VarName = ReadString(),
                    StartPc = ReadUint32(),
                    EndPc = ReadUint32()
                };
            }

            return locVars;
        }

        private uint[] ReadLineInfo()
        {
            var lineInfo = new uint[ReadUint32()];
            for (var i = 0; i < lineInfo.Length; i++)
            {
                lineInfo[i] = ReadUint32();
            }

            return lineInfo;
        }

        private Prototype[] ReadProtos(string parentSource)
        {
            var protos = new Prototype[ReadUint32()];
            for (var i = 0; i < protos.Length; i++)
            {
                protos[i] = new Prototype();
                protos[i] = ReadProto(parentSource);
            }

            return protos;
        }

        private Upvalue[] ReadUpvalues()
        {
            var upvalues = new Upvalue[ReadUint32()];
            for (var i = 0; i < upvalues.Length; i++)
            {
                upvalues[i] = new Upvalue
                {
                    Instack = ReadByte(),
                    Idx = ReadByte()
                };
            }

            return upvalues;
        }

        private object[] ReadConstants()
        {
            var constants = new object[ReadUint32()];
            for (var i = 0; i < constants.Length; i++)
            {
                constants[i] = ReadConstant();
            }

            return constants;
        }

        private object ReadConstant()
        {
            switch (ReadByte())
            {
                case BinaryChunk.TAG_NIL: return null;
                case BinaryChunk.TAG_BOOLEAN: return ReadByte() != 0;
                case BinaryChunk.TAG_INTEGER: return ReadLuaInteger();
                case BinaryChunk.TAG_NUMBER: return ReadLuaNumber();
                case BinaryChunk.TAG_SHORT_STR: return ReadString();
                case BinaryChunk.TAG_LONG_STR: return ReadString();
                default: throw new Exception("corrupted!");
            }
        }

        private uint[] ReadCode()
        {
            var code = new uint[ReadUint32()];
            for (var i = 0; i < code.Length; i++)
            {
                code[i] = ReadUint32();
            }

            return code;
        }

        private string ReadString()
        {
            var size = (uint) ReadByte() & 0xFF;
            if (size == 0)
            {
                return "";
            }

            if (size == 0xFF)
            {
                size = (uint) ReadUint64();
            }

            var bytes = ReadBytes(size - 1);
            return ConvertUtil.Bytes2String(bytes);
        }
    }
}